decisions/websre/0004-use-1-kubernetes-cluster-per-environment.md

```diff
@@ -0,0 +1,56 @@
+# 4. Use 1 Kubernetes Cluster Per Environment
+
+Date: 2021-06-09
+Scope: Web SRE Team
+
+## Status
+
+Proposed
+
+## Context
+
+We need to have multiple environments (sometimes called tiers) for Web SRE Services. These environments generally serve different purposes & audiences, and can have different uptime requirements. Environments can optionally serve as required points in the path towards the production environment. 
+
+At Mozilla, our service environments generally include Development (Dev), Staging (Stage), and Production (Prod). There can sometimes also be a Testing environment (usually ephemeral). Generally, these environments are in the same infrastructure provider account, and differ at the infrastructure or application platform level (e.g. at the Kubernetes level) - with the exception of Testing, which tends to be ephemeral within a particular context (e.g. a docker-compose spin up or a kind cluster in a CI pipeline).
+
+There are multiple ways that we differentiate these persistent (dev, stage, prod) environments on our infrastructure and application platforms. Within a Kubernetes context, some common patterns for separating persistent environments are:
+1. Having a distinct Kubernetes cluster per environment;
+2. Having one namespace per environment in a shared Kubernetes cluster;
+3. Having distinct node pool(s) per environment within a single cluster along with using selectors, taints, and tolerations.
+
+## Decision
+
+For Web SRE Services persistent environments - dev, stage, prod - that are Kubernetes-based, we will:
+* Use a distinct Kubernetes cluster per environment (dev, stage, prod) within the same infrastructure account;
+* Use a distinct Kubernetes namespace within each cluster per application;
+* Deploy applications to these distinct clusters as relevant to the expected application environment (e.g. applications should deploy to the stage cluster, prod applications should deploy to the prod cluster, etc.);
```

Comment:
Is this intending to say "e.g. stage applications"?

Reply:
yeah, for some reason in my head, that didn't sound right the first time. but looking at it again, that probably works better